### PR TITLE
ACQ-1589: fix payment terms display

### PR DESCRIPTION
**Pull request by ludrob**
### Description
Add support to display payment terms for offer's custom period of time (P6M, P3Y, etc) 

### Ticket
https://financialtimes.atlassian.net/browse/ACQ-1589

### Screenshots
| Before | After |
| ------ | ----- |
|image|image|
|image|image|
|image|image|

### Reminder
Have you completed these common tasks (remove those that don't apply)?

- [ ] **Documentation** updated or created
- [ ] **Tests** written for new or updated for existing functionality
- [ ] **Stories** updated to use this change
- [ ] **Accessibility** checked for screen readers and contrast
- [ ] **Design Review** ran past the designer
- [ ] **Product Review** ran past the product owner

### CODE REVIEW

File: components/payment-term.jsx [hunk 1]
```diff
@@ -160,14 +154,34 @@ export function PaymentTerm ({
 			);
 		};
 
-		const getDisplayName = () => {
-			let displayName = '';
+		const getTermDisplayName = () => {
```
Comment (ludrob) [inline components/payment-term.jsx:157]: I did only half the job last time, I should have move all the related code in there.
Let's try to tiday up all the paymentTerm display logic.

File: components/payment-term.jsx [hunk 2]
```diff
@@ -265,7 +351,7 @@ PaymentTerm.propTypes = {
 			selected: PropTypes.bool,
 			trialDuration: PropTypes.string,
 			trialPrice: PropTypes.string,
-			amount: PropTypes.number,
+			amount: PropTypes.string,
```
Comment (ekocian) [inline components/payment-term.jsx:354]: Response from membership is coming as a string

File: components/payment-term.jsx [hunk 3]
```diff
@@ -103,15 +159,9 @@ export function PaymentTerm ({
 			name: inputName,
 			value: option.value,
 			className:
-				'o-forms-input__radio o-forms-input__radio--right ncf__payment-term__input',
-			...(option.selected && { defaultChecked: true }),
+				'o-forms-input__radio o-forms-input__radio--right ncf__payment-term__input',			...(option.selected && { defaultChecked: true }),
```
Comment (maguero) [inline components/payment-term.jsx:162]: This line format looks weird... should it be in a new line?

File: components/payment-term.stories.js [hunk 4]
```diff
@@ -14,25 +14,80 @@ export const Basic = (args) => (
...
+			'discount': '',
+			'selected': false,
+			'price': '$909.00',
+			'trialPrice': '$0.00',
+			'trialDuration': '',
+			'monthlyPrice': '0',
+			'amount': '909.00',
+			'trialAmount': null,
+			'value': 'P3Y',
+			'currency': 'USD'
```
Comment (maguero) [inline components/payment-term.stories.js:46]: It seems there are different indent formats

**Review by maguero — APPROVED** [review]
Some minor comments about formats and indents....

File: components/payment-term.stories.js (hunk 4 above)
Comment (ekocian) [reply components/payment-term.stories.js:46]: b2029fd017d739119f6437d713e773149176662e

File: components/payment-term.jsx (hunk 3 above)
Comment (ekocian) [reply components/payment-term.jsx]: b2029fd017d739119f6437d713e773149176662e

**Review by ludrob — COMMENTED** [review]
Getting there, a few more comments.

Longer term: need to work with data sources to get all we need from there, display name, prices, etc.

File: components/payment-term.jsx [hunk 5]
```diff
@@ -143,31 +202,66 @@ export function PaymentTerm ({
...
-							</span>
-							{option.chargeOnText && (
-								<p className="ncf__payment-term__charge-on-text">
-									{option.chargeOnText}
-								</p>
-							)}
-						</div>
+						// this should cover the cases different than annual, quarterly and monthly
+						// for those containing period on option.value, render custom template, for the rest keep legacy render
+						typeof option.value === 'string' && !!(option.value.match(/^(P[0-9]+(M|Y))$/)) ?
```
Comment (ludrob) [inline components/payment-term.jsx:207]: more readable, less error prone
```suggestion
						typeof option.value === 'string' && Boolean(option.value.match(/^(P[0-9]+(M|Y))$/)) ?
```

File: components/payment-term.jsx [hunk 6]
```diff
@@ -143,31 +202,66 @@ export function PaymentTerm ({
...
 			}
-			const termPeriod = nameMap[option.name] ? title : option.title;
-			return `${displayName}${termPeriod} `;
+
+			const getTermPeriod = () => {
+				// annual, quarterly and monthly
+				if (nameMap[option.name]) {
+					return title;
+				// custom offer with period provided
+				} else if (typeof option.value === 'string' && !!(option.value.match(/^(P[0-9]+(M|Y))$/))) {
```
Comment (ludrob) [inline components/payment-term.jsx:254]: we return in the if statement, no need for else
```suggestion
				}
				if (typeof option.value === 'string' && !!(option.value.match(/^(P[0-9]+(M|Y))$/))) {
```

File: components/payment-term.jsx [hunk 7]
```diff
@@ -89,6 +134,37 @@ export function PaymentTerm ({
...
+			),
+			trialPrice: (trialPrice, trialPeriod) => (
+				<React.Fragment>
+					Unless you cancel during your trial you will be billed{' '}
+					<span className="ncf__payment-term__price">{trialPrice}</span> per {trialPeriod}
+					after the trial period.
+				</React.Fragment>
+			),
+			monthlyPrice: (monthlyPrice) =>
+				!!monthlyPrice && (
```
Comment (ludrob) [inline components/payment-term.jsx:153]:
```suggestion
				Boolean(monthlyPrice) && (
```

File: components/payment-term.jsx [hunk 8]
```diff
@@ -89,6 +134,37 @@ export function PaymentTerm ({
...
+					<span className="ncf__payment-term__equivalent-price">
+						That’s equivalent to{' '}
+						<span className="ncf__payment-term__monthly-price">{monthlyPrice}</span>{' '}
+						per month
+					</span>
+				),
+			renewsText: (renewalPeriod) => (
+				!!renewalPeriod && (
```
Comment (ludrob) [inline components/payment-term.jsx:161]:
```suggestion
				Boolean(renewalPeriod) && (
```

File: components/payment-term.jsx (hunk 6 above)
Comment (ekocian) [reply components/payment-term.jsx]: I have replaced that with some helper already present on n-pricing. please check if it works for you f2b7d53a3472b8f84348daefaabc8bb648984e72

File: components/payment-term.jsx (hunk 5 above)
Comment (ekocian) [reply components/payment-term.jsx]: https://github.com/Financial-Times/n-conversion-forms/pull/632#discussion_r870693249

File: components/payment-term.jsx (hunk 7 above)
Comment (ekocian) [reply components/payment-term.jsx]: 0711ab1035577294ab76993d2505d747012eca9d

File: components/payment-term.jsx (hunk 8 above)
Comment (ekocian) [reply components/payment-term.jsx]: 0711ab1035577294ab76993d2505d747012eca9d

File: components/payment-term.jsx [hunk 9]
```diff
@@ -143,31 +195,66 @@ export function PaymentTerm ({
...
-							</span>
-							{option.chargeOnText && (
-								<p className="ncf__payment-term__charge-on-text">
-									{option.chargeOnText}
-								</p>
-							)}
-						</div>
+						// this should cover the cases different than annual, quarterly and monthly
+						// for those containing period on option.value, render custom template, for the rest keep legacy render
+						typeof option.value === 'string' && !!(option.value.match(/^(P[0-9]+(M|Y))$/)) ?
```
[the remark this answers is not in the document]
Comment (ekocian) [reply components/payment-term.jsx]: https://github.com/Financial-Times/n-conversion-forms/pull/632#discussion_r870693249

File: components/payment-term.jsx [hunk 10]
```diff
@@ -8,11 +8,36 @@ export function PaymentTerm ({
...
+		const currencySymbol = price.match(/^.*?(?=[0-9])/);
+		const amount = price.match(/[0-9].*/);
+		const quantity = period.substring(1, period.length - 1);
+		const monthlyPrice = period.substring(period.length - 1) === 'Y' ?
+			(Number(amount) / (quantity * 12)).toFixed(2) :
+			(Number(amount) / quantity).toFixed(2);
+		return (
+			`${currencySymbol}${monthlyPrice}`
+		);
+	};
```
[the remark this answers is not in the document]
Comment (ekocian) [reply components/payment-term.jsx]: https://github.com/Financial-Times/n-conversion-forms/pull/632#discussion_r870693249

File: components/payment-term.jsx [hunk 11]
```diff
@@ -1,18 +1,63 @@
...
+		const monthlyPrice = periodObj.calculatePrice('P1M', amount);
+		return new Monthly({ value: monthlyPrice, currency }).getAmount('monthly');
+	}
+
+	/**
+	 * returns period converted to time if found
+	 * otherwise returns empty string to avoid show information not mapped
+	 * @param {string} period PxY (yearly) or PxM (montly) where x is the amount of years/months
+	 * @returns {string}
+	 *  */
```
Comment (ludrob) [inline components/payment-term.jsx:40]:
```suggestion
	 */
```

File: components/payment-term.jsx [hunk 12]
```diff
@@ -1,18 +1,63 @@
...
+	const getTimeFromPeriod = (period) => {
```
Comment (ekocian) [inline components/payment-term.jsx:41]: FYI: this is implemented with moment in next-subscribe. if we are going to remove it better leave as it is

File: components/payment-term.jsx [hunk 13]
```diff
@@ -143,31 +214,67 @@ export function PaymentTerm ({
...
 			}
-			const termPeriod = nameMap[option.name] ? title : option.title;
-			return `${displayName}${termPeriod} `;
+
+			const getTermPeriod = () => {
+				// annual, quarterly and monthly
+				if (nameMap[option.name]) {
+					return title;
+				// custom offer with period provided
+				} else if (isValidPeriod(option.value)) {
```
Comment (ludrob) [inline components/payment-term.jsx:266]:
```suggestion
				}
				if (isValidPeriod(option.value)) {
```

File: components/payment-term.jsx [hunk 14]
```diff
@@ -143,31 +214,67 @@ export function PaymentTerm ({
...
+					return getTimeFromPeriod(option.value);
+				}
+				// this covers custom legacy cases, where period is not provided
+				return option.title;
+
+			};
```
Comment (ludrob) [inline components/payment-term.jsx:272]:
```suggestion
				return option.title;
			};
```

**Review by AniaMakes — COMMENTED** [review]
I am going to brain-dump my thoughts here, they are more generic in nature than specific code lines. 

1️⃣  We need to make sure this is in line with whatever is home-brewed currently. For example `next-profile` has this https://github.com/Financial-Times/next-profile/blob/main/server/helpers/term.js. The page that uses it does not user conversion-forms. I guess the option is to make it a conversion forms component so that it can used a shared library and functionality. 
2️⃣  It looks like you're currently checking PxY and PxM, but I _think_ we've had instances using PxW and PxD in the past (we have them in the helper linked above, so I presume they were set up for a reason). 
3️⃣  did you know that this format is an ISO standard? https://www.digi.com/resources/documentation/digidocs/90001437-13/reference/r_iso_8601_duration_format.htm
4️⃣  I think we may have offer name mappers either in SDK or next-profile / next-retention - I'm guessing those will need to be looked at to check if everything is aligned as it should.

**Comment by ekocian** [comment]
> I am going to brain-dump my thoughts here, they are more generic in nature than specific code lines.
> 
> 1️⃣ We need to make sure this is in line with whatever is home-brewed currently. For example `next-profile` has this https://github.com/Financial-Times/next-profile/blob/main/server/helpers/term.js. The page that uses it does not user conversion-forms. I guess the option is to make it a conversion forms component so that it can used a shared library and functionality. 2️⃣ It looks like you're currently checking PxY and PxM, but I _think_ we've had instances using PxW and PxD in the past (we have them in the helper linked above, so I presume they were set up for a reason). 3️⃣ did you know that this format is an ISO standard? https://www.digi.com/resources/documentation/digidocs/90001437-13/reference/r_iso_8601_duration_format.htm 4️⃣ I think we may have offer name mappers either in SDK or next-profile / next-retention - I'm guessing those will need to be looked at to check if everything is aligned as it should.

I agree with most of your point of view @AniaMakes , actually we are thinking about to refactor this ugly code, but this is a temporary solution for an uncover case. But took more time and efforts than expected.
Regarding to ISO standard for period, I think jsdoc is quite confusing, we are using n-pricing for that validation, which covers the rest of the cases that you mention

Please @ludrob add whatever you think I'm missing

**Review by jorbuedo — APPROVED** [review]

**Review by ekocian — APPROVED** [review]

**Pull request merged**